V1 Prototype Snapshot

Olshansk opened the pull request:
This PR is meant to address this issue: https://github.com/pokt-network/pocket/issues/22.

It is a functional snapshot of the V1 prototype that can be referenced and used as a baseline as V1 development continues.

# Testing

## 0. Prerequisite

The development environment and instructions here are not expected to be complete since the development environment is still a work in progress, but the two following assumptions are made:
1. Docker is installed and the daemon is running locally.
2. `protoc` is installed and running locally.

## 1. Pull changes:
```
$ git clone git@github.com:pokt-network/pocket.git pocket_v1_proto_snapshot
$ cd pocket_v1_proto_snapshot
$ git checkout -b integration/workshop2/attempt2 origin/integration/workshop2/attempt2
$ cd prototype
```

## 2. Clean environment:
```
$ cd prototype
$ make docker_wipe # Press `y` if prompted
$ make v1_generate_protos # Note that this is not using the docker protos at the moment
```

## 3. Run localnet

Example video can be found [here](https://drive.google.com/file/d/1Uc43PQ4t15GyT5t1bpsv_KpCwyu9HxBa).

### 3.1 Shell 1:
```
$ make compose_and_watch
```

### 3.2 Shell 2:
```
$ make client_start
$ make client_connect
```

### 3.3. Run commands

In Shell:
1. Select `TogglePaceMakerManualMode`
2. Select `TogglePaceMakerManualMode`

luyzdeleon commented:
Shouldn't this PR be done to the integration issue branch? https://github.com/pokt-network/pocket/tree/milestone/v1-prototype

Also no reviewers, no milestone, no asignees and no labels on the PR

Olshansk commented:
@luyzdeleon I created the PR but it was still a WIP, so I should've marked it as a draft. Ready for review now!

Olshansk commented:
@andrewnguyen22  @derrandz Note that I tried to keep the codebase as close as possible to what was presented last Friday. The only changes I made to make the testing instructions above work are these:

```
497e559 Improve docker_wipe command
7c0d2f3 Updated Makefile to generate all the necessary protos
466be36 Removed 'build' directory from gitignore since it contains the Dockerfiles
63d26c0 Added helper to Makefile to remove all containers
4edb55b Removed root contribution guide
d8dad1d Merge branch 'main' into integration/workshop2/attempt2
8561ace Revert the README
a7078fd Reverted code in root directory since everything is in prototype
3013985 Added prototype README
3ee06f2 Moved all the prorotype code into a prototype directory
9c596be Fix Dockerfile paths
```

Olshansk commented:
Opened a PR here: https://github.com/pokt-network/pocket/pull/27

derrandz commented:
@Olshansk Good for me, I tried running this locally and everything works 👍🏻

derrandz commented on prototype/CONTRIBUTING.md:
```diff
@@ -0,0 +1,17 @@
+Reach out to us on [Telegram](https://t.me/POKTnetwork) and start a discussion with the team regarding your change before you start working. Communication is key for open source projects and asynchronous contributions.
+
+For an active research forum, checkout and post on [our forum](https://research.pokt.network).
+
```
I think discord can be used too, I think it's more practical than TG.

Olshansk replied:
Noted. This was just a transfer copy/paste of what we had before. Going to leave as is because we'll be deleting the `prototype` directory eventually and the newer instructions are in Notion.

derrandz commented on prototype/consensus/README.md:
```diff
@@ -0,0 +1,148 @@
+- [HotStuff: BFT Consensus in the Lens of Blockchain](https://arxiv.org/pdf/1803.05069.pdf)
+- [Twins: White-Glove Approach for BFT Testing](https://arxiv.org/pdf/2004.10617.pdf)
+- [Fast-HotStuff: A Fast and Robust BFT Protocol for Blockchains](https://arxiv.org/pdf/2010.11454.pdf)
+- [On the Performance of Pipelined HotStuff](https://arxiv.org/pdf/2107.04947.pdf)
```
This file is such an example. Looking forward to adding similar files to all modules.

Olshansk replied:
Noted. I was just using this as a measure to make some notes during my research.

derrandz commented on prototype/consensus/debug_handlers.go:
```diff
@@ -0,0 +1,111 @@
+	anyProto, err := anypb.New(consensusProtoMsg)
+	if err != nil {
+		m.nodeLogError("Error encoding any proto: " + err.Error())
+		panic("^")
```
While I was looking at this, it came to me that I also was thinking about what to do on panic behaviors, what is our logic for recovery and all of that... an epic on its own

Olshansk replied:
ACK. We'll design something like this correctly in the future.

Olshansk commented:
@derrandz Correct. The entire `prototype` directory is a way for us to take a snapshot of the prototype at the end of last week. Not code reviewed, but a reference we can diff against. This requires a proper code review though: https://github.com/pokt-network/pocket/pull/29.

@luyzdeleon @andrewnguyen22 PTAL (Please Take A Look) when you have a chance, just from a functionality perspective, because once we move this over into mainline via a handful of PRs, we'd be able to delete `prototype` altogether.

Olshansk commented:
@derrandz I was trying to imply that you SHOULD NOT do a thorough review of the code under `prototype`. The purpose of it is to have a reference in case something breaks. Thorough reviews should be done on other things moving forward.

iajrz commented:
Version of protoc should be specified.

andrewnguyen22 commented:
I am unable to complete this step:
```
gco -b integration/workshop2/attempt2 origin/integration/workshop2/attempt2
cd prototype
```
`-bash: gco: command not found`
Also tried amending: 
```
gcc -b integration/workshop2/attempt2 origin/integration/workshop2/attempt2
cd prototype
```
```
clang: error: unsupported option '-b integration/workshop2/attempt2'
clang: error: no such file or directory: 'origin/integration/workshop2/attempt2'
clang: error: no input files
```

andrewnguyen22 commented:
Functionality wise, seems good so approving

Olshansk commented:
@iajrz 

> Expected version of protoc should be specified.

I added a `TODO` in the ‘Makefile`: `# TODO(team): Tested locally with `protoc` version `libprotoc 3.19.4`. In the near future, only the Dockerfiles will be used to compile protos.`

@andrewnguyen22 

> gco: command not found

Sorry about that. I'll make sure to test with just plain bash in the future.

For context, it's an alias that's part of the ohmyzsh git plugin in case anyone is curious: https://github.com/ohmyzsh/ohmyzsh/blob/master/plugins/git/git.plugin.zsh#L111

> Failed at the proto generate step 2

Per the TODO I added above and since this is just prototype code, I will move forward with submitting it for now.

Olshansk commented:
@iajrz I accideently forgot to push the `protoc` related TODO before merging...

Instead of creating the overhead of a PR for a one-line comment, just FYI that I'm adding it as a commit in the [other PR](https://github.com/pokt-network/pocket/pull/29/commits/c05798c621b16ad56dbde2eee8be48400817266e).
